packages/jest-types/src/Circus.ts

```diff
@@ -31,9 +31,17 @@ export type Hook = {
   timeout: number | undefined | null;
 };
 
-export type EventHandler = (event: Event, state: State) => void;
+export type EventHandler = (event: Event, state: State) => void | Promise<void>;
```

Comment:
```suggestion
export type EventHandler =
  | ((event: SyncEvent, state: State) => void)
  | ((event: AsyncEvent, state: State) => Promise<void>);
```

Reply:
@SimenB, the suggested signature does not work well enough, leading to errors here:

![Screenshot from 2020-01-13 14-15-04](https://user-images.githubusercontent.com/1962469/72255346-2c11e100-360f-11ea-88e3-5516e8c7d7fa.png)

I think I understand your intention to make `EventHandler` type stricter, so hopefully, the version I have just committed is enough to your liking too:

```ts
export interface EventHandler {
  (event: AsyncEvent, state: State): void | Promise<void>;
  (event: SyncEvent, state: State): void;
}
```

I've left an ambiguous `void | Promise<void>` result for the `AsyncEvent` for the backward compatibility. See why:

If some project has already implemented an `EventHandler` for `jest-circus` returning `void`, IMO, this use case remains perfectly valid. We should not force people into rewriting the handlers necessarily to `async` functions, not to mention the radical distinction between sync and async events. Such a change would feel too breaking, to my taste.

The compromise `void | Promise<void>`  still might affect the existing codebase of someone, but that is fixable by a light extra typing. For instance, in `jest/packages/*` codebase, after the change, I had to fix signatures in those two places:

```diff
diff --git a/packages/jest-circus/src/__mocks__/testEventHandler.ts b/packages/jest-circus/src/__mocks__/testEventHandler.ts
index 23dfe2bed..ad948f8d6 100644
--- a/packages/jest-circus/src/__mocks__/testEventHandler.ts
+++ b/packages/jest-circus/src/__mocks__/testEventHandler.ts
@@ -7,7 +7,10 @@
 
 import {Circus} from '@jest/types';
 
-const testEventHandler: Circus.EventHandler = (event, state) => {
+const testEventHandler: Circus.EventHandler = (
+  event: Circus.Event,
+  state: Circus.State,
+) => {
   switch (event.name) {
     case 'start_describe_definition':
     case 'finish_describe_definition': {
diff --git a/packages/jest-circus/src/eventHandler.ts b/packages/jest-circus/src/eventHandler.ts
index 7b059983d..b947bc7d9 100644
--- a/packages/jest-circus/src/eventHandler.ts
+++ b/packages/jest-circus/src/eventHandler.ts
@@ -21,7 +21,10 @@ import {
   restoreGlobalErrorHandlers,
 } from './globalErrorHandlers';
 
-const eventHandler: Circus.EventHandler = (event, state): void => {
+const eventHandler: Circus.EventHandler = (
+  event: Circus.Event,
+  state: Circus.State,
+): void => {
```

That way or another, I mean just to warn you that going away from the former signature of mine:

```
export type EventHandler = (event: Event, state: State) => void | Promise<void>;
```

most likely will lead to guaranteed TypeScript errors in projects that already implement some `EventHandler`s on top of `jest-circus`. Please advise if this is something critical for decision-making here.